hazelcast/src/main/java/com/hazelcast/cache/impl/AbstractCacheService.java

```diff
@@ -385,32 +393,47 @@ public void deleteCache(String cacheNameWithPrefix, String callerUuid, boolean d
     public CacheConfig putCacheConfigIfAbsent(CacheConfig config) {
         // ensure all configs registered in CacheService are not PreJoinCacheConfig's
         CacheConfig cacheConfig = asCacheConfig(config);
-        CacheConfig localConfig = configs.putIfAbsent(cacheConfig.getNameWithPrefix(), cacheConfig);
-        if (localConfig == null) {
-            if (cacheConfig.isStatisticsEnabled()) {
-                setStatisticsEnabled(cacheConfig, cacheConfig.getNameWithPrefix(), true);
-            }
-            if (cacheConfig.isManagementEnabled()) {
-                setManagementEnabled(cacheConfig, cacheConfig.getNameWithPrefix(), true);
+        CacheConfigFuture future = new CacheConfigFuture(nodeEngine, cacheConfig);
+        CacheConfigFuture localConfigFuture = configs.putIfAbsent(cacheConfig.getNameWithPrefix(), future);
+        // if the existing cache config future is not yet fully configured, we block here
+        CacheConfig localConfig = localConfigFuture == null ? null : localConfigFuture.join();
+        if (localConfigFuture == null) {
+            try {
+                if (cacheConfig.isStatisticsEnabled()) {
+                    setStatisticsEnabled(cacheConfig, cacheConfig.getNameWithPrefix(), true);
+                }
+                if (cacheConfig.isManagementEnabled()) {
+                    setManagementEnabled(cacheConfig, cacheConfig.getNameWithPrefix(), true);
+                }
+                logger.info("Added cache config: " + cacheConfig);
+                additionalCacheConfigSetup(null, config);
+            } finally {
+                // now it is safe for others to obtain the new cache config
+                future.complete();
```

Comment:
I'm thinking what the error case should be here. This future is meant to provide ordering between cache config initialization and its usage. Once the caller of `getCacheConfig` gets a config later it assumes this initialization process succeeded. I see no behavior change here since the previous EE method overridden this was running after the `putIfAbsent` above. But I still see some gaps here. If an exception is thrown from `additionalCacheConfigSetup`, we can still get an NPE later when obtaining the WAN publisher for the cache.

Reply:
My intention was to maintain existing behavior, so even if exception is thrown from `additionalCacheConfigSetup`, the new `CacheConfig` would remain in `configs` map. This implementation is still safe from the NPE scenario in https://github.com/hazelcast/hazelcast-enterprise/issues/2877 because every concurrently executing thread will obtain the new `CacheConfig` and re-attempt `additionalCacheConfigSetup` (in line 415 below) -> all threads waiting on `putCacheConfigIfAbsent` will receive the exception.

However, you are right we might be susceptible to other faults. Maybe a better option would be to remove the new `CacheConfigFuture` from `configs` map in case of exception; this is cleaner and gives user a chance to provide a correct config in case of exception. So something like:

```
try {
  ...
  additionalCacheConfigSetup(null, config);
  // now it is safe for others to obtain the new cache config
  future.complete();
} catch (RuntimeException e) {
  future.complete(e);
  configs.remove(cacheConfig.getNameWithPrefix(), future);
  throw e;
}
```

wdyt?

Reply:
Yes, I was just thinking about removing from the map in the exception case, so :+1: 

I see the other callers will also receive the same exception. Missed that point.